File: linkpreviewedittext/src/main/java/in/codeshuffle/linkpreviewedittext/scraper/ParserUtils.java
package in.codeshuffle.linkpreviewedittext.scraper;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;

import in.codeshuffle.linkpreviewedittext.model.LinkInfo;
import in.codeshuffle.linkpreviewedittext.util.DocumentUtils;
import in.codeshuffle.linkpreviewedittext.util.Utils;

class ParserUtils {

    private final String originalUrl;

    ParserUtils(String originalUrl) {
        this.originalUrl = originalUrl;
    }

    /**
     * UnModular parse
     *
     * @param document document object
     */
    LinkInfo parseLinkDataUnModular(Document document) {
        LinkInfo linkInfo = new LinkInfo();

        Elements elements = document.getElementsByTag("meta");

        // getTitle document.select("meta[property=og:title]")
        String title = document.select("meta[property=og:title]").attr("content");

        if (title == null || title.isEmpty()) {
            title = document.title();
        }
        linkInfo.setTitle(title);

        //getDescription
        String description = document.select("meta[name=description]").attr("content");
        if (description == null || description.isEmpty()) {
            description = document.select("meta[name=Description]").attr("content");
        }
        if (description == null || description.isEmpty()) {
            description = document.select("meta[property=og:description]").attr("content");
        }
        if (description == null || description.isEmpty()) {
            description = "";
        }
        linkInfo.setDescription(description);


        // getMediaType
        Elements mediaTypes = document.select("meta[name=medium]");
        String type;
        if (mediaTypes.size() > 0) {
            String media = mediaTypes.attr("content");

            type = media.equals("image") ? "photo" : media;
        } else {
            type = document.select("meta[property=og:type]").attr("content");
        }
        linkInfo.setMediaType(type);


        //getImages
        Elements imageElements = document.select("meta[property=og:image]");
        if (imageElements.size() > 0) {
            String image = imageElements.attr("content");
            if (!image.isEmpty()) {
                linkInfo.setImageUrl(Utils.resolveURL(originalUrl, image));
            }
        }
        if (linkInfo.getImageUrl().isEmpty()) {
            String src = document.select("link[rel=image_src]").attr("href");
            if (!src.isEmpty()) {
                linkInfo.setImageUrl(Utils.resolveURL(originalUrl, src));
            } else {
                src = document.select("link[rel=apple-touch-icon]").attr("href");
                if (!src.isEmpty()) {
                    linkInfo.setImageUrl(Utils.resolveURL(originalUrl, src));
                    linkInfo.setFaviconUrl(Utils.resolveURL(originalUrl, src));
                } else {
                    src = document.select("link[rel=icon]").attr("href");
                    if (!src.isEmpty()) {
                        linkInfo.setImageUrl(Utils.resolveURL(originalUrl, src));
                        linkInfo.setFaviconUrl(Utils.resolveURL(originalUrl, src));
                    }
                }
            }
        }

        //Favicon
        String src = document.select("link[rel=apple-touch-icon]").attr("href");
        if (!src.isEmpty()) {
            linkInfo.setFaviconUrl(Utils.resolveURL(originalUrl, src));
        } else {
            src = document.select("link[rel=icon]").attr("href");
            if (!src.isEmpty()) {
                linkInfo.setFaviconUrl(Utils.resolveURL(originalUrl, src));
            }
        }

        for (Element element : elements) {
            if (element.hasAttr("property")) {
                String str_property = element.attr("property").trim();
                if (str_property.equals("og:originalUrl")) {
                    linkInfo.setUrl(element.attr("content"));
                }
                if (str_property.equals("og:site_name")) {
                    linkInfo.setSiteName(element.attr("content"));
                }
            }
        }

        if (linkInfo.getUrl().equals("") || linkInfo.getUrl().isEmpty()) {
            URI uri = null;
            try {
                uri = new URI(originalUrl);
            } catch (URISyntaxException e) {
                e.printStackTrace();
            }
            if (uri == null) {
                linkInfo.setUrl(originalUrl);
            } else {
                linkInfo.setUrl(uri.getHost());
            }
        }

        return linkInfo;
    }

    /**
     * Modular parse
     *
     * @param document document object
     */
    LinkInfo parseLinkDataModular(Document document) {
        LinkInfo linkInfo = new LinkInfo();

        //set the original url
        linkInfo.setUrl(originalUrl);

        //parse title
        String title = DocumentUtils.getLinkTitle(document);
        linkInfo.setTitle(title);

        //parse description
        String description = DocumentUtils.getDescription(document);
        linkInfo.setDescription(description);

        //parse favicon
        String faviconUrl = DocumentUtils.getFaviconUrl(document);
        linkInfo.setFaviconUrl(faviconUrl);

        //parse mediaType
        String type = DocumentUtils.getMediaType(document);
        linkInfo.setMediaType(type);


        //parse images
        String imageUrl = DocumentUtils.getImageUrl(originalUrl, document);
        linkInfo.setImageUrl(imageUrl);

        String domainUrl = DocumentUtils.getDomainUrl(originalUrl, document);
        linkInfo.setDomainUrl(domainUrl);

        String siteName = DocumentUtils.getUrlTitle(document);
        linkInfo.setSiteName(siteName);

        return linkInfo;
    }
}

File: linkpreviewedittext/src/main/java/in/codeshuffle/linkpreviewedittext/ParserUtils.kt
package `in`.codeshuffle.linkpreviewedittext

import org.jsoup.nodes.Document
import java.net.URI
import java.net.URISyntaxException

internal class ParserUtils(private val originalUrl: String) {

    /**
     * UnModular parse
     *
     * @param document document object
     */
    fun parseLinkDataUnModular(document: Document): LinkInfo {
        val linkInfo = LinkInfo()
        val elements = document.getElementsByTag("meta")

        // getTitle document.select("meta[property=og:title]")
        var title = document.select("meta[property=og:title]").attr("content")
        if (title == null || title.isEmpty()) {
            title = document.title()
        }
        linkInfo.title = title

        //getDescription
        var description = document.select("meta[name=description]").attr("content")
        if (description == null || description.isEmpty()) {
            description = document.select("meta[name=Description]").attr("content")
        }
        if (description == null || description.isEmpty()) {
            description = document.select("meta[property=og:description]").attr("content")
        }
        if (description == null || description.isEmpty()) {
            description = ""
        }
        linkInfo.description = description


        // getMediaType
        val mediaTypes = document.select("meta[name=medium]")
        val type: String
        type = if (mediaTypes.size > 0) {
            val media = mediaTypes.attr("content")
            if (media == "image") "photo" else media
        } else {
            document.select("meta[property=og:type]").attr("content")
        }
        linkInfo.mediaType = type


        //getImages
        val imageElements = document.select("meta[property=og:image]")
        if (imageElements.size > 0) {
            val image = imageElements.attr("content")
            if (!image.isEmpty()) {
                linkInfo.imageUrl = Utils.resolveURL(originalUrl, image)
            }
        }
        if (linkInfo.imageUrl?.isEmpty() == true) {
            var src = document.select("link[rel=image_src]").attr("href")
            if (!src.isEmpty()) {
                linkInfo.imageUrl = Utils.resolveURL(originalUrl, src)
            } else {
                src = document.select("link[rel=apple-touch-icon]").attr("href")
                if (!src.isEmpty()) {
                    linkInfo.imageUrl = Utils.resolveURL(originalUrl, src)
                    linkInfo.faviconUrl = Utils.resolveURL(originalUrl, src)
                } else {
                    src = document.select("link[rel=icon]").attr("href")
                    if (!src.isEmpty()) {
                        linkInfo.imageUrl = Utils.resolveURL(originalUrl, src)
                        linkInfo.faviconUrl = Utils.resolveURL(originalUrl, src)
                    }
                }
            }
        }

        //Favicon
        var src = document.select("link[rel=apple-touch-icon]").attr("href")
        if (!src.isEmpty()) {
            linkInfo.faviconUrl = Utils.resolveURL(originalUrl, src)
        } else {
            src = document.select("link[rel=icon]").attr("href")
            if (!src.isEmpty()) {
                linkInfo.faviconUrl = Utils.resolveURL(originalUrl, src)
            }
        }
        for (element in elements) {
            if (element.hasAttr("property")) {
                val str_property = element.attr("property").trim { it <= ' ' }
                if (str_property == "og:originalUrl") {
                    linkInfo.url = element.attr("content")
                }
                if (str_property == "og:site_name") {
                    linkInfo.siteName = element.attr("content")
                }
            }
        }
        if (linkInfo.url == "" || linkInfo.url?.isEmpty() == true) {
            var uri: URI? = null
            try {
                uri = URI(originalUrl)
            } catch (e: URISyntaxException) {
                e.printStackTrace()
            }
            if (uri == null) {
                linkInfo.url = originalUrl
            } else {
                linkInfo.url = uri.host
            }
        }
        return linkInfo
    }

    /**
     * Modular parse
     *
     * @param document document object
     */
    fun parseLinkDataModular(document: Document): LinkInfo {
        val linkInfo = LinkInfo()

        //set the original url
        linkInfo.url = originalUrl

        //parse title
        val title = DocumentUtils.getLinkTitle(document)
        linkInfo.title = Utils.getNonNullString(title)

        //parse description
        val description = DocumentUtils.getDescription(document)
        linkInfo.description = Utils.getNonNullString(description)

        //parse favicon
        val faviconUrl = DocumentUtils.getFaviconUrl(document)
        linkInfo.faviconUrl = Utils.getNonNullString(faviconUrl)

        //parse mediaType
        val type = DocumentUtils.getMediaType(document)
        linkInfo.mediaType = Utils.getNonNullString(type)


        //parse images
        val imageUrl = DocumentUtils.getImageUrl(originalUrl, document)
        linkInfo.imageUrl = Utils.getNonNullString(imageUrl)
        val domainUrl = DocumentUtils.getDomainUrl(originalUrl, document)
        linkInfo.domainUrl = Utils.getNonNullString(domainUrl)
        val siteName = DocumentUtils.getUrlTitle(document)
        linkInfo.siteName = Utils.getNonNullString(siteName)
        return linkInfo
    }

}